sympy/stats/joint_rv.py

```diff
@@ -0,0 +1,176 @@
+"""
+Joint Random Variables Module
+
+See Also
+========
+sympy.stats.rv
+sympy.stats.frv
+sympy.stats.crv
+sympy.stats.drv
+"""
+
+from __future__ import print_function, division
+
+__all__ = ['marginal_density']
+
+from sympy import Basic, Lambda, sympify, S
+from sympy.concrete.summations import Sum
+from sympy.integrals.integrals import Integral
+from sympy.stats.rv import (ProductPSpace, RandomSymbol,
+        NamedArgsMixin, ProductDomain)
+from sympy.core.containers import Tuple
+
+class JointPSpace(ProductPSpace):
+    """
+    Represents a joint probability space. Represented using symbols for
+    each component and a distribution.
+    """
+    def __new__(cls, syms, dist, *args):
+        syms = Tuple.fromiter(i for i in syms)
+        return Basic.__new__(cls, syms, dist, *args)
+
+    @property
+    def set(self):
+        return self.domain.set
+
+    @property
+    def symbols(self):
+        return self.args[0]
+
+    @property
+    def values(self):
+        vls = []
+        for sym in self.symbols:
+            vls.extend(MarginalPSpace((sym,), self).values)
+        return tuple(vls)
+
+    @property
+    def distribution(self):
+        return self.args[1]
+
+    def pdf(self, *args):
+        return self.distribution(*args)
+
+    def where(self, condition):
+        raise NotImplementedError()
+
+    def compute_density(self, expr):
+        raise NotImplementedError()
+
+    def sample(self):
+        raise NotImplementedError()
+
+    def probability(self, condition):
+        raise NotImplementedError()
+
+
+class MarginalPSpace(JointPSpace):
```

Comment:
what's the advantage of having this class separate from `JointPSpace`?

Reply:
Having it as a seperate class allows a different constructor, and each marginal probability space can  have a joint space associated with it. Also, creating a `JointPSpace` requires to specify a distribution, which is not needed for `MarginalPSpace`, which is identified on the basis of the joint probability space from which it is derived and the symbols in it.

Reply:
The marginal pspace object is not being returned anywhere right now. It's only being used internally to calculate the density. Should I add a function that allows the user to create and use the marginal probability space similar to a joint probability space?

Reply:
I'm a bit skeptical about inheriting from `ProductPSpace`: the way classes are designed in SymPy requires the arguments of a subclass constructor to match the arguments of the superclass constructor. You add `syms` as first arguments, which disrupts this logic.

Reply:
what about making `ProductPSpace` an abstract class and move the constructor to another subclass, like `IndependentProductPSpace` (this would manage the class for independent prob spaces)?

Reply:
>  Also, creating a JointPSpace requires specifying a distribution, which is not needed for MarginalPSpace; it can be identified on the basis of the joint probability space from which it is derived and the symbols in it.

If you change the constructor, then you should not inherit from `JointPSpace`. The marginal pspace is technically still a joint pspace. I see the reason for a separate class in the case you cannot calculate the marginal distribution, so you need to support the unevaluated expression, but I wouldn't spend too much time on this in this PR.

Reply:
Okay, I'll add a method for calculating the marginal density in `JointPSpace` and remove `MarginalPSpace` for now.